### PR TITLE
Improved subscription views

### LINKED ISSUE #63: Improve Subscription functionality
- Improve subscription model so foreign key will links to UserAndCompanyRelation instead of User
- Change authentication system in views

### PR DESCRIPTION
**Pull request by parsival2022**
- Rewritten subscription views for deleting, creating and retrieving.
- Retrieve serializer removed, replaced with inner model method for data formatting instead
- Companies views improved

### CODE REVIEW

**Comment by YanZhylavy** [comment]
Great job! The functionality is working generally correctly, I've checked, and now it's more integrated into the overall authentication logic in our project. However, I noticed a small issue - DRF serializers are not being used here. Please consider the possibility of fixing the old SubscriptionSerializer and connecting it to the view.

**Review by YanZhylavy — APPROVED** [review]

File: authentication/models.py [hunk 1]
```diff
@@ -98,26 +112,71 @@ class Company(models.Model):
     startup_idea = models.TextField(blank=True)
     tags = models.CharField(max_length=255, blank=True)
 
+    @classmethod
+    def get_company(cls, *args, **kwargs):
+        return cls.objects.get(**kwargs)
```
Comment (mehalyna) [inline authentication/models.py:116]: Be cautious with `.get()` as it raises `DoesNotExist` if no object is found and `MultipleObjectsReturned` if more than one object is found

File: authentication/models.py [hunk 2]
```diff
@@ -98,26 +112,71 @@ class Company(models.Model):
...
+    
+    def get_info(self):
+        fields = ('brand', 'common_info', 'contact_phone', 'contact_email')
+        startup_fields = ('product_info', 'startup_idea')
+        data = {}
+        company_type = self.get_company_type()
+        data['company_type'] = company_type
+        if company_type == STARTUP:
+            for k in startup_fields:
+                data[k] = self.__getattribute__(k)
```
Comment (mehalyna) [inline authentication/models.py]: What about potential `AttributeError` ?

File: authentication/models.py [hunk 3]
```diff
@@ -98,26 +112,71 @@ class Company(models.Model):
...
+        for k in fields:
+                data[k] = self.__getattribute__(k)
+        return data
```
Comment (mehalyna) [inline authentication/models.py]: **DRY**  `get_info()` method contains some repetitive code. Consider ways to streamline this, possibly by defining a method or utility function for fetching attribute values

File: authentication/views.py [hunk 4]
```diff
@@ -53,11 +54,24 @@ def post(self, request):
...
+class LogoutView(APIView):
+    authentication_classes = (UserAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request):
+        try:
+            refresh_token = request.data["refresh_token"]
+            token = RefreshToken(refresh_token)
+            token.blacklist()
+            return Response(status=status.HTTP_205_RESET_CONTENT)
```
Comment (mehalyna) [inline authentication/views.py]: Using `HTTP 205 Reset Content `as the response status code is an interesting choice. While it indicates that the user agent should reset the document view, which might make sense in the context of a logout, `HTTP 204 No Content` is more commonly used for logout endpoints. `HTTP 204` indicates that the server has successfully fulfilled the request and that there is no additional content to send in the response payload body

File: authentication/views.py [hunk 5]
```diff
@@ -53,11 +54,24 @@ def post(self, request):
...
+        except Exception as e:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
```
Comment (mehalyna) [inline authentication/views.py:85]: Catching the base `Exception` class is generally not recommended as it can catch unexpected errors, making debugging difficult. It's better to catch specific exceptions. In the context of blacklisting a refresh token, you might specifically catch `TokenError` from `rest_framework_simplejwt.exceptions`

File: companies/models.py [hunk 6]
```diff
@@ -1,9 +1,23 @@
 from django.db import models
+from authentication.models import CompanyAndUserRelation, Company
 
-from authentication.models import CustomUser, Company
 class Subscription(models.Model):
-    subscription_id = models.BigAutoField(primary_key=True)
-    investor = models.ForeignKey(CustomUser, on_delete=models.CASCADE)
+    subscription_id = models.BigAutoField(primary_key=True, unique=True)
```
Comment (mehalyna) [inline companies/models.py:5]: The `unique=True` attribute is redundant here since `primary_key=True` already enforces uniqueness

File: companies/views.py [hunk 7]
```diff
@@ -1,86 +1,83 @@
...
+            print(serializer)
+            company = Company(**serializer.data)
+            company.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class CompaniesRetrieveUpdateView(APIView):
+    permission_classes = (IsAuthenticated)
```
Comment (mehalyna) [inline companies/views.py]: Ensure `permission_classes` is a tuple by adding a trailing comma when there's only one item: `(IsAuthenticated,)`

File: companies/views.py [hunk 8]
```diff
@@ -1,86 +1,83 @@
...
+    def get(request, pk=None):
+        if pk:
+            company = Company.get_company(company_id=pk)
+            return Response(company.get_info(), status=status.HTTP_200_OK)
+        type = request.data.get('company_type')
+        if not type:
+           companies = Company.get_all_companies_info()
+        else:
+            companies = Company.get_all_companies_info(type)
+        return Response(companies, status=status.HTTP_200_OK)
```
Comment (mehalyna) [inline companies/views.py]: The `get()` method should be an instance method with `self` parameter: `def get(self, request, pk=None):`. Without `self`, Django won't recognize this as a valid view method

Comment (mehalyna) [inline companies/views.py]: The view handles both individual company retrieval and listing companies based on type. Consider splitting these functionalities into separate views or DRF's `ViewSet` or `GenericAPIView` and mixins for clarity and according to **SRP**

File: companies/views.py [hunk 9]
```diff
@@ -1,86 +1,83 @@
...
+                        'company':Company.get_company(company_id=company_id)}
+            except (Company.DoesNotExist, CompanyAndUserRelation.DoesNotExist) as e:
+                return Response({'error': str(e)}, status=status.HTTP_404_NOT_FOUND)
+            serializer = SubscriptionSerializer(data=data)
+            if serializer.is_valid():
+                subscription = Subscription(**data)
+                subscription.save()
+                msg = f"You're successfully subscribed to {subscription.company.brand}"
+                return Response({'message': msg, 'subscription_id': subscription.subscription_id}, status=status.HTTP_201_CREATED)
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (mehalyna) [inline companies/views.py]: The nested `try-except` blocks make the code a bit hard to follow. Consider simplifying the logic or using Django's `get_or_create()` method to streamline subscription creation

File: companies/views.py [hunk 10]
```diff
@@ -1,86 +1,83 @@
...
+        except Subscription.DoesNotExist:
+            try:
+                data = {'investor': CompanyAndUserRelation.get_relation(relation_id=profile_id), 
+                        'company':Company.get_company(company_id=company_id)}
+            except (Company.DoesNotExist, CompanyAndUserRelation.DoesNotExist) as e:
+                return Response({'error': str(e)}, status=status.HTTP_404_NOT_FOUND)
+            serializer = SubscriptionSerializer(data=data)
+            if serializer.is_valid():
+                subscription = Subscription(**data)
+                subscription.save()
```
Comment (mehalyna) [inline companies/views.py]: Instead of manually creating and saving the `Subscription` instance, consider using `serializer.save()` after validating the serializer. This approach leverages DRF's serialization and validation mechanisms https://www.django-rest-framework.org/api-guide/serializers/

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: authentication/models.py (hunk 1 above)
Comment (parsival2022) [reply authentication/models.py:116]: This exceptions should be handled from outside since our model methods basically just encapsulate objects default methods. Since we may need different approach to exception handling, the exceptions aren't handled inside this methods.

File: authentication/models.py (hunk 2 above)
Comment (parsival2022) [reply authentication/models.py]: Since this method isn't classmethod, we assume that the instance of the class has all the attributes; even if the required field is blank, the instance would return None. Nevertheless, it's really useful to have additional insurance. I have added AttributeError handling so now, if it raised, None will be returned as value of the attribute in question.

File: authentication/models.py (hunk 3 above)
Comment (parsival2022) [reply authentication/models.py]: Agree, get_attribute method created

File: authentication/views.py (hunk 4 above)
Comment (parsival2022) [reply authentication/views.py]: Agree, status of the response changed to 204.

File: authentication/views.py (hunk 5 above)
Comment (parsival2022) [reply authentication/views.py:85]: Agree, Exception changed to TokenError.

File: companies/models.py (hunk 6 above)
Comment (parsival2022) [reply companies/models.py:5]: Agree, unique argument removed.
Sorry for this craving for uniqueness.)

File: companies/views.py (hunk 7 above)
Comment (parsival2022) [reply companies/views.py]: Comma was added

File: companies/views.py (hunk 8 above)
Comment (parsival2022) [reply companies/views.py]: Argument added.

Comment (parsival2022) [reply companies/views.py]: Agree, the view was splitted, additionally the urls was changed.

File: companies/views.py (hunk 9 above)
Comment (parsival2022) [reply companies/views.py]: Agree, code was simplified.

File: companies/views.py (hunk 10 above)
Comment (parsival2022) [reply companies/views.py]: Agree, fixed.

**Review by mehalyna — APPROVED** [review]

File: authentication/views.py [hunk 11]
```diff
@@ -93,7 +107,7 @@ class UserUpdateView(generics.RetrieveUpdateAPIView):
     permission_classes = (IsAuthenticated, CustomUserUpdatePermission)
 
 
-class UserPasswordUpdateView(generics.UpdateAPIView):
+class UserPasswordUpdateView(generics.UpdateAPIView): #doesn't this view double the password reset and recovery functions?
     queryset = CustomUser.objects.all()
     serializer_class = UserPasswordUpdateSerializer
     authentication_classes = (UserAuthentication,)
```
Comment (akostromin14) [inline authentication/views.py:113]: We have to discuss it at the meeting today.

File: forum/errors.py [hunk 12]
```diff
@@ -1,69 +1,90 @@
...
     """Pre-formatted errors for easier handling"""
 
-    class NO_HEADER:
+    class BaseError:
+        msg = None
+        status = None
+
+        @classmethod
+        def response(cls):
+            return Response({'error': cls.msg}, status=cls.status)
```
Comment (akostromin14) [inline forum/errors.py:14]: This is a good rule, but now we need to update this response everywhere in the project.

**Review by akostromin14 — APPROVED** [review]
Good PR! Why do you delete everything in the search app?

File: forum/errors.py (hunk 12 above)
Comment (parsival2022) [reply forum/errors.py:14]: Well, it's not neccessery, because you can still get attributes from the Error class directly or not use them at all. I just thought that it would be better to return standart form of the error response authomatically instead of write it manually again and again.

**Review by VovaD199 — APPROVED** [review]
